Post/smaller the parameter smaller the bugs (#11)
rakutentech/engineering-blog: content/post/smaller-the-parameter-smaller-the-bugs.md

---
title: "Smaller the Parameter, Smaller the Bugs"
date: 2021-07-13T22:57:39+09:00
hero: /images/hero-5.jpg

> excerpt: "Some hints for coding efficiently and reducing the bugs"

Reviewer: ```suggestion
excerpt: "Some tips for coding efficiently and reducing bugs"
```

timeToRead: 10
authors:
  - Bater
draft: true
---

> When I review the code, I sometimes notice that there is some room to squash from the parameter of the method, but not many developers are aware of it when they design the code. According to my coding experience, we can write the code efficiently and reduce the bug rate by the effort of reducing the parameter from the method. Unless your manager still evaluates your output by the line of code every day, you should definitely try these tips in your daily coding life.

Reviewer: [grammar] Unnecessary "the" articles.
```suggestion
When I review code, I sometimes notice that there is room to squash parameters of methods, yet, not many developers are aware of this when they design code. From my coding experience, we can write code efficiently and reduce bug rates by putting in effort in reducing the parameters from our methods. Unless your manager still evaluates output by lines of code, you should definitely try these tips in your daily coding life.
```

> So today, I will share how to do it with OOP (object-oriented programming) design, and explain where the magic power comes from.

Reviewer: ```suggestion
So today, I will share how to do it with OOP (object-oriented programming) design, and explain where this magic power comes from.
```

> Let's start with a small example, the demo code was written in Ruby, but don't worry. The ruby code is so friendly that once you have some OOP background, I believe you can understand the code right away. 

Reviewer: ```suggestion
Let's start with a small example. The following demo code was written in Ruby, but don't worry; Ruby code is so beginner-friendly that once you have some OOP background, it's easy to grasp.
```

What is the good code? This question might be hard to answer, but I would say **the good code should be small and simple.** In this article, let's focus on making the parameter smaller.

Generally speaking, **one parameter is better than two, and no parameter is better than one.** The fact of the parameter is, it cost time to read and understand the business logic, and time is always expensive in all projects. We spent around 70% of the time reading the existing code, but only 30% is actually writing. If we can save some time from reading the code, not only from literally reading but also from understanding the spec and logic, which means we can develop efficiently.

Secondly, when **the code is short, the bugs are nowhere to hide.** Parameter causes bugs easily because it needs extra concern. Reducing parameters can make the method more clear, and show the intention explicitly. It can help us to maintain the code easily. When we try to reduce the parameter, we are actually trying to reduce the potential bugs at the same time.

Now we know it's better to reduce the parameter, but how to do it? Let's check the parameter again.

## Data clump

> When we find out the parameters are **a group of data that always come together**, and it would make no sense when missing one of them, we can call it a set of **Data Clump**. For example, a pair of start date and end date, or in this example, first name and last name.

Reviewer: ```suggestion
One observation that we can make is that parameters are **a group of data that always comes together**. This data would make no sense when one piece of data is missing. We call it a set of **Data Clump**. For example, a start date and end date pair we would call a data clump or, as in this example, a data clump would be the first name and last name.
```

> In the OOP world, the **object is the container of data, and class is the container of method** (behavior of object). It's cheap to create many small objects to manage and operate the serial data. Additionally, we can define the method on the object and reuse it when necessary.

Reviewer: [grammar] shouldn't need a `,` when you join just two clauses with `and`

```suggestion
In the OOP world, the **object is the container of data and class is the container of methods** (i.e. the behavior of objects). It's cheap to create many small objects to manage and operate on serial data. Additionally, we can define methods on the object and reuse them when necessary.
```

Author: > I think the right approach is a footnote explaining that:
> 
> 1. Many small objects could end up causing GC pressure.
> 2. But high performance languages often have tricks up their sleeve like doingg [escape analysis](https://shipilev.net/jvm/anatomy-quarks/18-scalar-replacement/) to convert heap allocation to stack allocations or [expanding argument objects as primitive arguments](https://shipilev.net/jvm/anatomy-quarks/18-scalar-replacement/). These languages also generally have separate GC pathways for small objects and short-lived objects. In most cases, if you're just using an argument object locally in a high performance language, your runtime will manage to optimize most of the cost away.

Hi @boazy , thanks for your feedback.

The link for `escape analysis ` and `expanding argument objects as primitive arguments ` are the same, is that your intention?

end
```
The cover method generates a string that has the book title and author's full name inside. After the refactoring, we can reuse the author's full name method again. Once we build a good pattern, the new code can follow it easily and keep the project clean. But if you didn't pay attention, the code getting rust rapidly and becomes the developer's nightmare. 

You may also notice that both `summary` and `cover` need the same parameter, author, it's the smell of refactoring. The author information should be one of the attributes of the book object, instead of passing as a parameter when we call the book method. So we can refactor the book class again.

```rb
class Book
  attr_reader :title, :publish_year, :author
  
  def initialize(title, publish_year, author)
    @title = title
    @publish_year = publish_year
    @author = author
  end
 
  def summary
    "#{title} - #{author.full_name} (#{publish_year})"
  end
  
  def cover
    "#{title}\n\n\n#{author.full_name}"
  end
end
  
bater = Author.new("Bater", "Chen")
book = Book.new("Ruby is the best", "2020", bater)
book.summary
=> "Ruby is the best - Chen, Bater (2020)"
book.cover
=> "Ruby is the best\n\n\nChen, Bater"
```

> Both `summary` and `cover` method's output remain the same, but now we can call it freely without any parameter once the book object is prepared. In ruby, we can skip parentheses if there is no parameter.

Reviewer: ```suggestion
Both `summary` and `cover` methods' output remain the same but now we can call it freely without any parameter once the book object is prepared. In Ruby, we can skip parentheses when there is no parameter.
```

This example doesn't encourage you to move all the parameters into the object attributes, it depends on the domain know-how. We can do it because it makes sense for a book object to having an author as an attribute in this case. It may get worse if we try to put the wrong data as the object attribute.

## Open-closed principle
Reducing the parameter from two to zero is not the only benefit we got from this refactoring, this new design is open for extension but closed for modification. It follows the [open-closed principle][3].

> Now we have a new requirement coming again, some authors have the middle name. In the new version, it would be very easy to adapt to the change, we don't need to touch the book class anymore. The new author class would refactor like this:

Reviewer: ```suggestion
Let's say we have a new requirement again, this time, some authors now have a middle name. In this new version, it would be very easy to adapt our code because we don't need to touch the book class anymore. The new author class would refactored like this:
```

  
robert = Author.new("Robert", "Kiyosaki", "T")
Book.new("Rich Dad, Poor Dad", "1997", robert).summary
=> "Rich Dad, Poor Dad - Kiyosaki, T, Robert (1997)"
```
In this design, the existing behavior doesn't change. The default value of the middle name is `nil` (Null) because in most cases we can skip it, and we won't break anything when the existing code tries to create a new author with only two parameters. The author class is open for extension and the book class is closed for modification in this case. Without this refactoring, we need to touch on both `summary` and `cover` methods, and the parameters increase from two to three. Let's compare the code before and after the refactoring again:
#### Old pattern without refactoring
```rb
def summary(first_name, last_name, middle_name = nil)
  if middle_name.nil?
    "#{title} - #{last_name}, #{first_name} (#{publish_year})"
  else
    "#{title} - #{last_name}, #{middle_name}, #{first_name} (#{publish_year})"
  end
end
```
#### After refactoring
```rb
def summary
  "#{title} - #{author.full_name} (#{publish_year})"
end
```
Obviously, the code quality has a huge improvement after refactoring. 
## Loose coupling

> Before we finish this refactoring and send the PR to someone, let's check the code again, does it clean enough?

Reviewer: ```suggestion
Before we finish refactoring and send off this PR to someone, let's check the code again. But is it clean enough?
```

```rb
def summary
  # 10 lines of code you don't want to read.
  middle_result = "#{some_author_title_helper(author)} - #{author.full_name}"
  # another 10 lines of code
  return final_result
end
```
The disadvantage of coupling is that when one class needs to change, you need to change others as well. For example, one day we need to rename the full name method or adding more parameters to it, we need to modify the book class at the same time, which we don't like.

The bad news is, in the real world it's almost impossible to remove all the coupling between each object, but we can at least reduce the impact by segregate and highlight the dependence. 

```rb
def summary
  "#{title} - #{author_name} (#{publish_year})"
end

def author_name
  author.full_name
end
```
After this refactoring, the `summary` method 100% calling from the book object itself. The book class still has the coupling with the author class, but it's obvious and easy to maintain. If we only consider the example case, you might wonder is that a little bit over design? Well, I would say that a good pattern is never too early. Now we can push the PR confidently and enjoy the holiday.

## Conclusion

> The good code should be small and simple, of course, the parameter is not the only place we can do our effort to reduce the size. During the refactoring, don't forget to confirm the result by unit test to make sure the final behavior would be the same. People around you may think that it would be time-wasting work because the final behavior remains the same even though developers spent a lot of time on it. But there's the value of code quality. I believe it would perform in development speed and product stability in the long term.

Reviewer: ```suggestion
Good code should be small and simple. Of course, the parameter is not the only place we can place our effort in to reducing its size. During refactoring, don't forget to confirm the result by unit testing to make sure the final behavior is consistent. People around you may think that it would be time-wasting work because the final behavior remains the same even though developers spent a lot of time on it. But there's the value of code quality. I believe this line of thinking would help you perform well in development speed and product stability in the long term.
```

> I hope this small example can convince you this refactoring skill can benefit your coding life. If not, please waiting for my next blog and give me another chance, see you next time.

Reviewer: ```suggestion
I hope this small example can convince you this refactoring skill can benefit your coding life. If not, please wait for my next blog and give me another chance. See you next time.
```

So today, I will share how to do it with OOP (object-oriented programming) design, and explain where this magic power comes from.

Let's start with a small example. The following demo code was written in Ruby, but don't worry; Ruby code is so beginner-friendly that once you have some OOP background, it's easy to grasp.

The example story is, we have some books and want to pass the author name as a parameter, and get the book title, author name, and publish date as a summary.
```ruby
class Book
  attr_reader :title, :publish_year
 
  def initialize(title, publish_year)
    @title = title
    @publish_year = publish_year
  end
 
  def summary(first_name, last_name)
    "#{title} - #{last_name}, #{first_name} (#{publish_year})"
  end
end
 
book = Book.new("Ruby is the best", "2020")
book.summary("Bater", "Chen")
=> "Ruby is the best - Chen, Bater (2020)"
```

> We have a book class, it has two attributes, title, and publish_year. The `initialize` method would be called when we `new` a book object. Book class only has one instance method called `summary`, it needs two parameters, first_name and last_name. In Ruby, when we define the `attr_reader` as the attribute of the object, it will create the read method automatically. So we can call the instance variable directly by `title` to get `@title` return, same as publish_year.

Reviewer: Maybe "title" and "publish_year" could be formatted with enclosing \`\`? e.g.  it has two attributes: `title` and `publish_year`. I see a bunch of other places where you're referencing classes and object names without \`\`

```rb
class Author
  attr_reader :first_name, :last_name
 
  def initialize(first_name, last_name)
    @first_name = first_name
    @last_name = last_name
  end
 
  def full_name
    "#{last_name}, #{first_name}"
  end
end
```
The Author class also has two attributes, and the author object has one instance method call `full_name`. Now we can use the author object to replace the first name and last name.

#### Before
```rb
def summary(first_name, last_name)
  "#{title} - #{last_name}, #{first_name} (#{publish_year})"
end
 
book = Book.new("Ruby is the best", "2020")
book.summary("Bater", "Chen")
=> "Ruby is the best - Chen, Bater (2020)"
```
#### After
```rb
def summary(author)
  "#{title} - #{author.full_name} (#{publish_year})"
end
 
book = Book.new("Ruby is the best", "2020")
book.summary(Author.new("Bater", "Chen"))
=> "Ruby is the best - Chen, Bater (2020)"
```
The output of `summary` is the same, but the parameter reduces from two strings to one object. Well, you may doubt that it doesn't really matter, but imagine that you can follow the same pattern to reduce the parameter from 10 data into 1 object. That would be a huge impact.

## Law of Demeter

> You may also notice that I relocate the logic of full name from book class to author class because a full name can perform independently without interaction with a book object. An **Object should know less about each other**, it is known as the [Law of Demeter][2] or "Least Knowledge Principle".

Reviewer: ```suggestion
You may also have noticed that I relocated the logic of full name from book class to author class because a full name can perform independently without interaction with the book object. An **object should know less about each other** is said by the [Law of Demeter][2] or the "Least Knowledge Principle".
```

This practice is also known as [Duck Typing][1]. If it walks like a duck and it quacks like a duck, then it must be a duck. In duck typing, an object's suitability is determined by the presence of certain methods and properties, rather than the type of the object itself. In the Java case, you can use the interface.

Now a new requirement is coming, we need to add a new method called `cover`, and it also has author first name and last name inside. We can follow the pattern we just made. Let's compare before and after the refactoring of the book class.
#### Old pattern without refactoring
```rb
def summary(first_name, last_name)
  "#{title} - #{last_name}, #{first_name} (#{publish_year})"
end
 
def cover(first_name, last_name)
  "#{title}\n\n\n#{last_name}, #{first_name}"
end
```
#### After refactor
```rb
def summary(author)
  "#{title} - #{author.full_name} (#{publish_year})"
end
 
def cover(author)
  "#{title}\n\n\n#{author.full_name}"
end
```

> The cover method generates a string that has the book title and author's full name inside. After the refactoring, we can reuse the author's full name method again. Once we build a good pattern, the new code can follow it easily and keep the project clean. But if you didn't pay attention, the code would get rust rapidly and becomes the developer's nightmare. 

Reviewer: ```suggestion
The cover method generates a string that has the book title and author's full name inside. After the refactoring, we can reuse the author's full name method again. Once we build up a good pattern, the new code can follow it easier and keep our project clean. But if you didn't pay attention, the code would quickly become harder to read, develop  technical debt and become a developer's worst nightmare. 
```
Wasn't sure how to correct "code get rusty rapidly". You might need to reword this

```rb
class Book
  attr_reader :title, :publish_year, :author
  
  def initialize(title, publish_year, author)
    @title = title
    @publish_year = publish_year
    @author = author
  end
 
  def summary
    "#{title} - #{author.full_name} (#{publish_year})"
  end
  
  def cover
    "#{title}\n\n\n#{author.full_name}"
  end
end
  
bater = Author.new("Bater", "Chen")
book = Book.new("Ruby is the best", "2020", bater)
book.summary
=> "Ruby is the best - Chen, Bater (2020)"
book.cover
=> "Ruby is the best\n\n\nChen, Bater"
```
Both `summary` and `cover` method's output remain the same, but now we can call it freely without any parameter once the book object is prepared. In ruby, we can skip parentheses if there is no parameter.

This example doesn't encourage you to move all the parameters into the object attributes, it depends on the domain know-how. We can do it because it makes sense for a book object to having an author as an attribute in this case. It may get worse if we try to put the wrong data as the object attribute.

## Open-closed principle

> Reducing the parameter from two to zero is not the only benefit we got from this refactoring, this new design is open for extension but closed for modification. It follows the [open-closed principle][3].

Reviewer: ```suggestion
When we reducing our parameters from two to zero, our design is now open for extension but closed for modification. This is inline with the [open-closed principle][3].
```

Now we have a new requirement coming again, some authors have the middle name. In the new version, it would be very easy to adapt to the change, we don't need to touch the book class anymore. The new author class would refactor like this:

```rb
class Author
  attr_reader :first_name, :last_name, :middle_name
  
  def initialize(first_name, last_name, middle_name = nil)
    @first_name = first_name
    @last_name = last_name
    @middle_name = middle_name
  end
  
  def full_name
    if middle_name.nil?
      "#{last_name}, #{first_name}"
    else
      "#{last_name}, #{middle_name}, #{first_name}"
    end
  end
end
  
bater = Author.new("Bater", "Chen")
Book.new("Ruby is the best", "2020", bater).summary
=> "Ruby is the best - Chen, Bater (2020)"
  
robert = Author.new("Robert", "Kiyosaki", "T")
Book.new("Rich Dad, Poor Dad", "1997", robert).summary
=> "Rich Dad, Poor Dad - Kiyosaki, T, Robert (1997)"
```

> In this design, the existing behavior doesn't change. The default value of the middle name is `nil` (Null) because in most cases we can skip it, and we won't break anything when the existing code tries to create a new author with only two parameters. The author class is open for extension and the book class is closed for modification in this case. Without this refactoring, we need to touch on both `summary` and `cover` methods, and the parameters increase from two to three. Let's compare the code before and after the refactoring again:

Reviewer: ```suggestion
In this design, the existing behavior doesn't change. The default value of the middle name is `nil` (also known as Null in other languages) because in most cases we can skip it and we won't break anything when the existing code tries to create a new author with only two parameters. The author class is open for extension and the book class is closed for modification in this case. Without refactoring, we need to touch upon both the `summary` and `cover` methods, causing the parameters to increase from two to three. Let's compare the code before and after, again:
```

## Conclusion
Good code should be small and simple. Of course, the parameter is not the only place we can place our effort in to reducing its size. During refactoring, don't forget to confirm the result by unit testing to make sure the final behavior is consistent. People around you may think that it would be time-wasting work because the final behavior remains the same even though developers spent a lot of time on it. But there's the value of code quality. I believe this line of thinking would help you perform well in development speed and product stability in the long term.

I hope this small example can convince you this refactoring skill can benefit your coding life. If not, please wait for my next blog and give me another chance. See you next time.

[1]: https://en.wikipedia.org/wiki/Duck_typing
[2]: https://en.wikipedia.org/wiki/Law_of_Demeter

> [3]: https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle

Reviewer 2: I don't have time to review all the grammar but it needs more proofreading.

The content itself is good and generally important. Some general comments:

1. We should probably mention this is a particular case of a more general code-smell called [Primitive Obsession](https://refactoring.guru/smells/primitive-obsession). Primitive Obsession was coined by Martin Fowler, as far as I can tell.
2. We should also mention the [Whole Value idiom](http://fit.c2.com/wiki.cgi?WholeValue) promoted by Ward Cunningham even earlier. This provides the philosophical reasoning behind "why we merge first name and last name" - because these things represents a different value.
3. The `Author` example is not great IMHO. The actual value we really have in this argument is a person's _name_, not an author. A proper `Author` class would have things like birthday, country, biography and profile image. I think a better example would be for a `PersonName` which can be reused in parameters to functions, the Author class and even the User class. Another example could be the canonical [Whole Value](http://fit.c2.com/wiki.cgi?WholeValue): A pair of currency type and value (`15.33` and `:USD`).
